tool/src/c2/ty.rs

```diff
@@ -29,20 +34,33 @@ impl<'tcx> super::CContext<'tcx> {
         }
 
         // In some cases like generating decls for `self` parameters,
-        // a header will get its own forwards and includes. Instead of
+        // a header will get its own includes. Instead of
         // trying to avoid pushing them, it's cleaner to just pull them out
         // once done
-        let ty_name = context.cx.formatter.fmt_type_name(id);
-        context.header.forwards.remove(&*ty_name);
-        context.header.includes.remove(&*header_name);
+        context.decl_header.includes.remove(&*decl_header_path);
+        context.impl_header.includes.remove(&*impl_header_path);
 
-        self.files.add_file(header_path, header.to_string());
+        context
```

Comment:
suggestion: since the decl header is "special" ought we avoid it being sorted in and instead `#include` it specially at the end? i.e. the first line of the impl's body should be to include the decl header.